### PR TITLE
[FIX] base: prevent deleting a company if it has children

### PR DESCRIPTION
**Pull request by ushyme**
### Steps to reproduce

* create the following hierarchy of companies: 
  ```
    Company
    └── Branch A
        └── Branch B
  ```
* delete `Branch A`

From then on, whatever you do will result in a Bad Request response.

opw-3662711

### CODE REVIEW

**Comment by mart-e** [comment]
Have you tried fixing the root cause instead?
After a quick check, it seems that the problem is when computing `session_info` because `parent_path` is not correctly recomputed after deleting the company B.

My guess is that this is not limited to res.company but any model with a parent_path and no ondelete=cascade (but maybe there is only res.company in that case).

Something like 
```py
    def unlink(self):
        res = super().unlink()
        self._parent_store_compute()
        return res
```
should fix it but to check with @odoo/rd-framework-py if worth a solution in the ORM

**Comment by ushyme** [comment]
Thanks for taking the time to look into this.

You're right that the problem probably extends to any model with a `parent_path` and without `ondelete=cascade`.
However, in the case of `res.company`, there are a couple of additional considerations:
* deleting a company, if at all possible, tends to lead to significant complications
* modifying the company hierarchy is explicitly [not allowed](https://github.com/odoo/odoo/blob/5c05d8179ba757fc0b3067c8297e8636adc882e5/odoo/addons/base/models/res_company.py#L297-L298), likely due to concerns related to accounting.

Because of this, I think that the best immediate solution for this specific issue is to prevent the user from deleting a parent company, and check with the ORM team to see if it's worth fixing the general issue.

**Comment by ryv-odoo** [comment]
Indeed, there is a issue in the ORM side. We never fix the parent_path of `record`'s children after `record.unlink()` (it is only  true for parent field with self.ondelete='set null', which is the default AFAIK). IMO, it should be fixed, I created a task for it (3724457).

@ushyme , for company issue, why not adding  `ondelete='restrict'` on the parent_id field and let PostgreSQL manage that. Or I miss something ?

**Comment by ryv-odoo** [comment]
Hello @ushyme 

> The main reason why I chose not to use `ondelete='restrict'` is that it would require a module update, which means it wouldn't really fix the issue for existing databases.

Fair enough. But it should be the fix for master at least (adding `ondelete='restrict'`) if it is ok with you?

File: odoo/addons/base/models/res_company.py [hunk 1]
```diff
@@ -283,6 +283,11 @@ def cache_invalidation_fields(self):
             'sequence', # user._get_company_ids and other potential cached search
         }
 
+    @api.ondelete(at_uninstall=False)
+    def _unlink_if_company_has_no_childre(self):
```
Comment (ryv-odoo) [inline odoo/addons/base/models/res_company.py:287]: `children` maybe ?

File: odoo/addons/base/models/res_company.py [hunk 2]
```diff
@@ -283,6 +283,11 @@ def cache_invalidation_fields(self):
...
+        if self.child_ids:
```
Comment (ryv-odoo) [inline odoo/addons/base/models/res_company.py:288]: Careful, it can be call in batched.

File: odoo/addons/base/tests/test_res_company.py [hunk 3]
```diff
@@ -50,3 +50,17 @@ def test_logo_check(self):
...
+        """Ensure that companies with child companies cannot be deleted."""
+
+        parent_company = self.env['res.company'].create({
+            'name': 'Parent Company',
+            'child_ids': [
+                Command.create({'name': 'Child Company'}),
+            ],
+        })
+
+        with self.assertRaisesRegex(UserError, 'Companies that have associated branches cannot be deleted'):
```
Comment (ryv-odoo) [inline odoo/addons/base/tests/test_res_company.py:64]: Not need to check the content of the error. Moreover it will easier for the FW in master.

**Review by ryv-odoo — APPROVED** [review]
Small comments about the code itself but I am ok with the idea.

Thank you @ushyme

File: odoo/addons/base/models/res_company.py (hunk 2 above)
Comment (ushyme) [reply odoo/addons/base/models/res_company.py]: oops, this is embarrassing :sweat_smile:

**Comment by ryv-odoo** [comment]
@robodoo delegate+

**Comment by ushyme** [comment]
@robodoo r+

**Pull request closed without merging**